Text controls: Intrinsic widths should not be smaller than averageCharWidth() * n

Ahmad-S792 opened the pull request:
#### fae1b87b037b33725fc665867d88024c20b02f5f
<pre>
Text controls: Intrinsic widths should not be smaller than averageCharWidth() * n
<a href="https://bugs.webkit.org/show_bug.cgi?id=264897">https://bugs.webkit.org/show_bug.cgi?id=264897</a>
<a href="https://rdar.apple.com/118727006">rdar://118727006</a>

Reviewed by NOBODY (OOPS!).

This patch is to align WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/src/+/a00094b9b640c10fff554af6dcb4a68d78300321">https://chromium.googlesource.com/chromium/src/+/a00094b9b640c10fff554af6dcb4a68d78300321</a>

RenderTextControl::getAverageCharWidth(), which is the source of intrinsic
widths of INPUT/TEXTAREA calls for FontCascade::fastAverageCharWidthIfAvailable().
It should return a maximum value of &apos;width&apos; and rounded &apos;width&apos;.

In this patch, we removed `roundf` from FontCascade::fastAverageCharWidthIfAvailable()
to RenderTextControl::getAverageCharWidth() for more control and also leverage `font`
local variable throughout function. From previous attempt feedback, this also tackled
to keep the change in `RenderTextControl`.

* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::fastAverageCharWidthIfAvailable const): Remove `roundf` here
* Source/WebCore/rendering/RenderTextControl.cpp:
(WebCore::RenderTextControl::getAverageCharWidth): Move `rounding` here for control
* LayoutTests/fast/forms/text-control-intrinsic-widths.html: Rebaselined
* LayoutTests/platform/ios/fast/forms/text-control-intrinsic-widths-expected.txt: Ditto
* LayoutTests/platform/ios/fast/forms/textarea-width-expected.txt: Ditto
* LayoutTests/platform/mac/fast/forms/text-control-intrinsic-widths-expected.txt: Ditto
* LayoutTests/platform/mac/fast/forms/textarea-width-expected.txt: Ditto
* LayoutTests/platform/ios/fast/forms/text-control-intrinsic-widths-expected.txt: Ditto
* LayoutTests/platform/mac-ventura/fast/forms/text-control-intrinsic-widths-expected.txt: Ditto
* LayoutTests/fast/forms/textarea/textarea-cols-lucida-console.html: Add Test Case
* LayoutTests/fast/forms/textarea/textarea-cols-lucida-console-expected.txt: Add Test Case Expectation
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fae1b87b037b33725fc665867d88024c20b02f5f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82705 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2355 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36865 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87824 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33763 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84798 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2427 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10267 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64438 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22203 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85761 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1785 "Found 4 new test failures: fast/forms/auto-fill-button/input-strong-password-auto-fill-button.html fast/forms/auto-fill-button/input-strong-password-viewable.html fast/forms/search-styled.html fast/forms/textarea-width.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75231 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44713 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1667 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-relevancy-updates.html imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea.html (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29417 "Found 6 new test failures: fast/forms/auto-fill-button/input-strong-password-auto-fill-button.html fast/forms/auto-fill-button/input-strong-password-viewable.html fast/forms/search-styled.html fast/forms/text-control-intrinsic-widths.html fast/forms/textarea-width.html imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32796 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72939 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30080 "Found 6 new test failures: fast/forms/auto-fill-button/input-strong-password-auto-fill-button.html fast/forms/auto-fill-button/input-strong-password-viewable.html fast/forms/search-styled.html fast/forms/text-control-intrinsic-widths.html fast/forms/textarea-width.html imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89191 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9997 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7249 "Found 7 new test failures: fast/forms/auto-fill-button/input-strong-password-auto-fill-button.html fast/forms/auto-fill-button/input-strong-password-viewable.html fast/forms/search-styled.html fast/forms/text-control-intrinsic-widths.html fast/forms/textarea-width.html http/wpt/mediarecorder/record-96KHz-sources.html imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72855 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10225 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71040 "Exiting early after 10 failures. 107 tests run. 1 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72069 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16224 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15127 "Found 6 new test failures: fast/forms/auto-fill-button/input-strong-password-auto-fill-button.html fast/forms/auto-fill-button/input-strong-password-viewable.html fast/forms/search-styled.html fast/forms/text-control-intrinsic-widths.html fast/forms/textarea-width.html imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea.html (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1386 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9950 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15471 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9824 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13289 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11593 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderTextControl.cpp:
```diff
@@ -138,12 +139,18 @@ void RenderTextControl::hitInnerTextElement(HitTestResult& result, const LayoutP
+        // * Removing roundf() would make the intrinsic width smaller, and it
+        //   would have a compatibility risk.
+        return std::max(width, roundf(width));
+    }
```
This function should return LayoutUnit and  use LayoutUnit's ceiling. It would make it consistent with the rest of the intrinsic size ceiling you find across the codebase.

Ahmad-S792 replied:
@alanbaradlay - is this something you want? Using `LayoutUnit::fromFloatCeil`?

alanbaradlay replied:
yes, but with returning LayoutUnit type.

Ahmad-S792 committed:
Text controls: Intrinsic widths should not be smaller than averageCharWidth() * n

https://bugs.webkit.org/show_bug.cgi?id=264897
rdar://118727006

Reviewed by NOBODY (OOPS!).

This patch is to align WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/src/+/a00094b9b640c10fff554af6dcb4a68d78300321

RenderTextControl::getAverageCharWidth(), which is the source of intrinsic
widths of INPUT/TEXTAREA calls for FontCascade::fastAverageCharWidthIfAvailable().
It should return a maximum value of 'width' and rounded 'width'.

In this patch, we removed `roundf` from FontCascade::fastAverageCharWidthIfAvailable()
to RenderTextControl::getAverageCharWidth() for more control and also leverage `font`
local variable throughout function. From previous attempt feedback, this also tackled
to keep the change in `RenderTextControl`.

* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::fastAverageCharWidthIfAvailable const): Remove `roundf` here
* Source/WebCore/rendering/RenderTextControl.cpp:
(WebCore::RenderTextControl::getAverageCharWidth): Move `rounding` here for control
* LayoutTests/fast/forms/text-control-intrinsic-widths.html: Rebaselined
* LayoutTests/platform/ios/fast/forms/text-control-intrinsic-widths-expected.txt: Ditto
* LayoutTests/platform/ios/fast/forms/textarea-width-expected.txt: Ditto
* LayoutTests/platform/mac/fast/forms/text-control-intrinsic-widths-expected.txt: Ditto
* LayoutTests/platform/mac/fast/forms/textarea-width-expected.txt: Ditto
* LayoutTests/platform/ios/fast/forms/text-control-intrinsic-widths-expected.txt: Ditto
* LayoutTests/platform/mac-ventura/fast/forms/text-control-intrinsic-widths-expected.txt: Ditto
* LayoutTests/fast/forms/textarea/textarea-cols-lucida-console.html: Add Test Case
* LayoutTests/fast/forms/textarea/textarea-cols-lucida-console-expected.txt: Add Test Case Expectation
